# "Occupancy maps of 208 chromatin-associated proteins in one human cell type"

## All the codes used in this project, and study, are segregated into their own language specific platforms : Python, R and Shell (awk,Bash)

##### `Code execution may require installation of following softwares and dependencies:`

```- Python 2.7
   - Python >= 3.5
   - R >= 3.5
   - Shell Env
   - Bowtie2
   - spp
   - idr
   - meme-chip
   - Centrimo
   - TomTom
   - Bedtools
   - Fimo
   - Trimmomatic
   - Fastqc
   - Deeptools
   - kmer-svm
   - gkm-svm
   - Pybedtools
   - pandas
   - numpy
   - scipy
   - scikit-learn
   - dplyr
   - ranger
   - ComplexHeatmap
   - igraph
```

##### `Databases used in the study:`

```
   cisbp
   jaspar

```

##### `Inputs for the study are chiefly TFBS(Transcription factor binding sites) as a peak bed files mapped by ChIP-Seq and CETCh-Seq assay`


```
   Some analysis implements the standalone scripts, and some the combination of multiple or chaining scripts, including pipes.

   Codes built here were not originally intended for packages, but serves as a framework, and maybe considered 

   in future for more extensive analysis purpose, as required.
```

`Note : Please cite the orginal software and program if used in the study, for example, SPP, IDR, Bedtools, meme-suite and kmer/gkm-SVM... based papers`

